### ISSUE TITLE
[MPS] Nondeterministic NaN/Inf gradients born in backward() when training torchvision Faster R-CNN (finite loss, non-finite grads); single-stage FCOS/RetinaNet unaffected

### ISSUE THREAD

**Issue by vishalmahulkar** [pytorch/pytorch#190632]
### 🐛 Describe the bug

Training torchvision's `fasterrcnn_mobilenet_v3_large_fpn` on MPS nondeterministically produces non-finite gradients that are **born in `backward()`**: the loss is finite, but after `loss.backward()` many parameter grads are NaN/Inf. The same batch with the same initial weights is sometimes clean and sometimes not across trials, which points to a race in an MPS backward kernel rather than a numerical/config problem. CPU is always clean on identical inputs.

Self-contained repro (synthetic data, no dataset needed):

```python
import torch
import torchvision


def make_batch(seed=0, n_imgs=4, hw=(1080, 1920)):
    g = torch.Generator().manual_seed(seed)
    imgs = [torch.rand(3, *hw, generator=g) for _ in range(n_imgs)]
    tgts = []
    for _ in range(n_imgs):
        n = int(torch.randint(1, 6, (1,), generator=g))
        xy = torch.rand(n, 2, generator=g) * torch.tensor([1800.0, 1000.0])
        wh = torch.rand(n, 2, generator=g) * torch.tensor([120.0, 60.0]) + 8
        tgts.append({"boxes": torch.cat([xy, xy + wh], 1),
                     "labels": torch.randint(1, 25, (n,), generator=g)})
    return imgs, tgts


def trial(device, steps=5):
    """Short training loop; returns (step, loss, bad_param_names) at first non-finite grad."""
    torch.manual_seed(0)
    model = torchvision.models.detection.fasterrcnn_mobilenet_v3_large_fpn(
        weights=None, weights_backbone="DEFAULT", num_classes=25).to(device).train()
    opt = torch.optim.SGD(model.parameters(), lr=1e-4, momentum=0.9)
    imgs_cpu, tgts_cpu = make_batch()
    imgs = [i.to(device) for i in imgs_cpu]
    tgts = [{k: v.to(device) for k, v in t.items()} for t in tgts_cpu]
    for s in range(steps):
        losses = model(imgs, tgts)
        loss = sum(losses.values())
        opt.zero_grad()
        loss.backward()
        loss_val = float(loss.detach())
        bad = [n for n, p in model.named_parameters()
               if p.grad is not None and not torch.isfinite(p.grad).all()]
        if not torch.isfinite(loss) or bad:
            return s, loss_val, bad
        opt.step()
    return None, loss_val, []


if __name__ == "__main__":
    print(f"torch {torch.__version__} torchvision {torchvision.__version__}")
    s, loss, bad = trial(torch.device("cpu"), steps=2)
    print(f"cpu control: first_bad_step={s} loss={loss:.4f} bad={len(bad)}")
    for t in range(4):
        s, loss, bad = trial(torch.device("mps"))
        print(f"mps trial {t}: first_bad_step={s} loss={loss:.4f} "
              f"bad_grad_params={len(bad)}" + (f" first={bad[:3]}" if bad else ""),
              flush=True)
```

Output on my machine (M1 Max, 64 GB, macOS 26.3):

```
torch 2.13.0 torchvision 0.28.0
cpu control: first_bad_step=None loss=3.8955 bad=0
mps trial 0: first_bad_step=2 loss=nan bad_grad_params=50 first=['backbone.body.7.block.0.0.weight', 'backbone.body.7.block.1.0.weight', 'backbone.body.7.block.2.0.weight']
mps trial 1: first_bad_step=0 loss=3.9824 bad_grad_params=54 first=['backbone.body.7.block.0.0.weight', 'backbone.body.7.block.1.0.weight', 'backbone.body.7.block.2.0.weight']
mps trial 2: first_bad_step=0 loss=3.9824 bad_grad_params=54 first=['backbone.body.7.block.0.0.weight', 'backbone.body.7.block.1.0.weight', 'backbone.body.7.block.2.0.weight']
mps trial 3: first_bad_step=0 loss=3.9824 bad_grad_params=54 first=['backbone.body.7.block.0.0.weight', 'backbone.body.7.block.1.0.weight', 'backbone.body.7.block.2.0.weight']
```

Note trials 1–3: **loss is finite (3.98), yet 54 param grads are non-finite immediately after backward on step 0** — the NaNs originate inside backward, not from a diverging forward. Repro frequency varies run to run (it's a race): in an isolated session I measured NaN on 2/3 trials with an identical captured real-data batch + fixed weights (clean on the 3rd); with other GPU work running concurrently it reproduced 4/4. CPU on the very same batch/weights is always finite. I also have a 176 MB captured batch + model state from real training that replays the failure, available on request.

What I ruled out while debugging (all on the same batch/weights):

- **Not data / not divergence** — CPU is always clean; failures happen at lr=1e-4 and under warmup (lr≈2e-6) alike; loss is finite when grads are not.
- **Not the usual suspects in isolation** — backbone-only forward/backward: clean; `roi_align` at production scale: clean; `nms`: clean; `interpolate`: clean; disabling in-place activations: no change.
- **Architecture-correlated** — single-stage models with the same resnet50-FPN backbone (`fcos_resnet50_fpn`, `retinanet_resnet50_fpn`) train stably on MPS for 300+ batches (finite losses, real learning; I have since trained FCOS for many epochs on MPS without an issue). The failing configurations are the RoI-head (Faster R-CNN) models, and the first non-finite grads consistently land on the earliest trainable params of the failing chain (e.g. `backbone.body.7.*` above), i.e. the accumulation point deepest in backward.

Possibly related: `fasterrcnn_resnet50_fpn_v2` on MPS doesn't NaN for me — instead it hard-hangs at 0% CPU in what looks like the Metal command-queue deadlock pattern of #144634. Happy to file that separately if it isn't the same root cause.

### Versions

```
PyTorch version: 2.13.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 26.3 (arm64)
GCC version: Could not collect
Clang version: 17.0.0 (clang-1700.6.4.2)
CMake version: Could not collect
Libc version: N/A

Python version: 3.14.5 (main, May 10 2026, 10:21:34) [Clang 21.0.0 (clang-2100.0.123.102)] (64-bit runtime)
Python platform: macOS-26.3-arm64-arm-64bit-Mach-O
Is CUDA available: False
Is XPU available: False
Is XNNPACK available: False

CPU: Apple M1 Max

Versions of relevant libraries:
[pip3] numpy==2.5.1
[pip3] torch==2.13.0
[pip3] torchmetrics==1.9.0
[pip3] torchvision==0.28.0
```

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93 @mruberry @kurtamohler @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @Isalia20 (module: mps)

**Comment by Isalia20** [comment]
let me take a look, thanks for the report

**Comment by Isalia20** [comment]
This is a torchvision issue which has been fixed. Can you test it out with the nightly version of torch and torchvision and see if it works?
PR that fixed it
https://github.com/pytorch/vision/pull/9510

**Comment by vishalmahulkar** [comment]
Root-caused and fixed: this was the torchvision `roi_align` MPS **backward** kernel over-accumulating input gradients (grid-stride loop strided by a single threadgroup's width), fixed in pytorch/vision#9510 (merged 2026-07-15).

Verified on nightly `torch 2.14.0.dev20260721` / `torchvision 0.29.0.dev20260721` on the same machine (M1 Max, macOS 26.3):

- The repro script from this issue: **4/4 MPS trials clean** (finite gradients, loss decreasing) vs 4/4 failing on stable 2.13/0.28.
- The related `fasterrcnn_resnet50_fpn_v2` Metal command-queue deadlock mentioned above is **also gone** (5 training steps complete with finite decreasing loss) — evidently downstream of the same kernel bug.

The over-accumulation factor growing with RoI count also explains why my isolated roi_align test had looked clean — it ran below the RoI count where the corruption becomes catastrophic.

Closing as fixed by pytorch/vision#9510.
